LwM2m JSON format impl

alseddnm opened the pull request:
This is our 1st attempt to make a major contribution in Leshan, let us know if our code does not meet your conventions and standards, we are very flexible to make any modifications.

jvermillard commented on leshan-core/pom.xml:
```diff
@@ -38,7 +38,10 @@ Contributors:
             <artifactId>gson</artifactId>
         </dependency>
-
+		<dependency>
```
formating ? ;(

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java:
```diff
@@ -169,6 +169,10 @@ private ResourceModel parseResource(Node item) {
                 case "Time":
                     type = Type.TIME;
                     break;
+                case "Objlnk":
```
IMO adding Object link should be part of another pull request, it's a different feature

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java:
```diff
@@ -109,6 +122,13 @@ public static LwM2mNode decode(byte[] content, ContentFormat format, LwM2mPath p
             }
             return new LwM2mResource(path.getResourceId(), Value.newBinaryValue(content));
         case JSON:
+        	try {
```
formating?

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java:
```diff
@@ -109,6 +122,13 @@ public static LwM2mNode decode(byte[] content, ContentFormat format, LwM2mPath p
+				LwM2mJsonObject json =  LwM2mJson.fromJsonLwM2m(jsonStrValue);
+				return parseJSON(json,path, model);
+			} catch (LwM2mJsonException e) {
+			    throw new InvalidValueException("Unable to deSerialize json.", path, e);
```
remove '.'  at the end of the exception message

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java:
```diff
@@ -210,7 +247,83 @@ private static LwM2mNode parseTlv(Tlv[] tlvs, LwM2mPath path, LwM2mModel model)
+           // TODO 
+           // If bn is present will have multiple object instances in JSON the payload
+           // In case of JSON contains ObjLnk -> this should return List<LwM2mNode> ???
+        	 throw new InvalidValueException("Not yet implemented: " + " Object Path ", path);
```
throw new InvalidStateException("Not implemented")

alseddnm replied:
Where to find InvalidStateException?
I don't think is part of the JDK, do you mean IllegalStateException?

alseddnm replied:
I can also Implement InvalidStateException.java, But it will be used only for this purpose?

alseddnm replied:
InvalidStateException not part of the JDK and Leshan don't have it. Do you mean to implement it?

alseddnm replied:
@jvermillard ???

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeEncoder.java:
```diff
@@ -85,7 +90,13 @@
+            node.accept(jsonEncoder);
+            encoded = jsonEncoder.encoded;
+          break;
+//            throw new IllegalArgumentException("JSON content format not supported");
```
remove this line

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java:
```diff
@@ -0,0 +1,174 @@
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = -7863478129361403779L;
```
please remove , if you really want it use version 1

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	@Override
+	public String toString() {
```
StringFormat please

alseddnm replied:
You mean to use String.format() instead of StringBuilder?

alseddnm replied:
@jvermillard ???

jvermillard replied:
look at the other toString in the Leshan code, String.format 
look at https://github.com/jvermillard/leshan/wiki/Code-%26-design-guidelines

alseddnm replied:
@jvermillard  Yes I did read the guidelines, However toString has conditional statements in this case, I don't think is good to have an empty string printed in toString output?
The value could be float, String, Boolean, string or objectLink. Only one value will be displayed in the to Sting output. The time value also can be empty.
I have seen StringBuilder has been used  in Leshan Core (ResourceModel.java, ObjectModel.java,...) ..
Let me know what you think?

Thanks

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java:
```diff
@@ -0,0 +1,68 @@
+	
+    /**
+     * <ul> This <code>toJson</code> method accepts LwM2mJsonObject object and converts to JSON String.
+	 * <li>
```
?

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java:
```diff
@@ -0,0 +1,68 @@
+			return element;
+		} catch (Exception e) {
+			LOG.error(" Unable to deserialize JSON String to LwM2mJsonElement object: " +e.getMessage());
+			throw new LwM2mJsonException(
```
log or throw, you need to choose :)

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java:
```diff
@@ -0,0 +1,108 @@
+
+public class LwM2mJsonObject implements Serializable{
+	
+	private static final long serialVersionUID = 7302201912160751832L;
```
please use version 1

alseddnm replied:
done.

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java:
```diff
@@ -0,0 +1,108 @@
+
+	@Override
+	public String toString() {
+		return "LwM2mJsonElement [baseName=" + baseName + ", baseTime="
```
StringFormat

alseddnm replied:
done.

jvermillard commented on leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java:
```diff
@@ -152,4 +167,53 @@ public void tlv_power_source__multiple_resource() throws InvalidValueException {
+    public void json_device_object_instance0() throws InvalidValueException {
+        // tlv content for instance 0 of device object
+    	String content = "{"
+    			+ "\"e\":["+"{"+"\"n\":\"0\","+"\"sv\":\"Open Mobile Alliance\""+"},"
```
please remove the unused "+"

alseddnm replied:
done.

alseddnm commented:
@jvermillard We have all the changes in our fork JSON format branch, how do we create a new pull request. Should we close this one 1st?
Sorry we kind of learning GitHub process.
Thank you,
Nizar...!

sophokles73 commented:
Hi Nizar,

you do not necessarily create a new pull request since GitHub automatically adds your subsequent changes to the existing PR. However, some of the authors do not seem to have a valid CLA in place and in some commits the SIgned-off-by footer does not match the author of the commit :-(
From taking a look at the [issues GitHub complained about](https://dev.eclipse.org/eclipse-webhook/services/status_details.php?id=5554cde6d1335) it looks like you used two different email addresses to log into GitHub ...

Kai

alseddnm commented:
@sophokles73 I have corrected the email address issue, changed it in my GitHub and Eclipse account to use my gmail. As of this pull request should I do any action or just leave it as is?

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java:
```diff
@@ -169,7 +169,7 @@ private ResourceModel parseResource(Node item) {
                     type = Type.TIME;
                     break;
-                }
+                 }
```
WTF?

alseddnm replied:
@jvermillard  what you mean by WTF?
We removed the OBJECTLINK, as you said it has to be in a different PR.

jvermillard replied:
you just added a space char to this file

alseddnm replied:
removed ..

jvermillard commented on leshan-core/pom.xml:
```diff
@@ -1,93 +1,86 @@
-<!--
```
Why you reformatted this file?

alseddnm replied:
sorry I will revert this change..

jvermillard commented on leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonException.java:
```diff
@@ -0,0 +1,31 @@
+        super(message, cause);
+    }
+
+    private static final long serialVersionUID = 5765213465835546341L;
```
version 1

jvermillard commented on leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java:
```diff
@@ -15,7 +15,9 @@
  *******************************************************************************/
 package org.eclipse.leshan.core.node.codec;
 
-import static org.junit.Assert.*;
```
keep the static import at it was

alseddnm commented:
@jvermillard  This looks OK to you?
Thanks,
Nizar..!

jvermillard commented:
Can you fix this? : 

    The following users do not have valid CLA. Please see this page for details, or to sign a CLA.
        nalsedde@gemalto.com
        nalsedde@dvn8bs1.gemalto.com

    The following users have invalid Signed-off-by footers. Please use the Signed-Off-By (git commit -s) mechanism to confirm the origin of the submission. The identity in the Signed-off-By must match the Author or the Committer.
        nalsedde@gemalto.com
        nalsedde@dvn8bs1.gemalto.com
